File: common-lib/src/com/bigcorp/common/rules/RuleContext.java
package com.bigcorp.common.rules;

import com.bigcorp.common.model.TradeOrder;
import com.bigcorp.common.model.Client;
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

/**
 * Context object passed through the rule chain.
 * Contains the order being evaluated plus any additional data
 * rules might need.
 * 
 * Also accumulates messages and warnings from rules.
 * 
 * @author Bob
 * @since 1.0
 */
public class RuleContext {

    private TradeOrder order;
    private Client client;
    private Map attributes;  // generic key-value store for rule data
    private List messages;    // rule evaluation messages (strings)
    private List warnings;    // warnings that don't block the order
    private boolean rejected;
    private String rejectionReason;

    public RuleContext(TradeOrder order, Client client) {
        this.order = order;
        this.client = client;
        this.attributes = new HashMap();
        this.messages = new ArrayList();
        this.warnings = new ArrayList();
        this.rejected = false;
    }

    public TradeOrder getOrder() {
        return order;
    }

    public Client getClient() {
        return client;
    }

    public void setAttribute(String key, Object value) {
        attributes.put(key, value);
    }

    public Object getAttribute(String key) {
        return attributes.get(key);
    }

    public void addMessage(String message) {
        messages.add(message);
    }

    public List getMessages() {
        return messages;
    }

    public void addWarning(String warning) {
        warnings.add(warning);
    }

    public List getWarnings() {
        return warnings;
    }

    public boolean isRejected() {
        return rejected;
    }

    public void reject(String reason) {
        this.rejected = true;
        this.rejectionReason = reason;
    }

    public String getRejectionReason() {
        return rejectionReason;
    }
}

File: rewrite-experiment/hypothesis-a/src/domain/RuleContext.ts
import { TradeOrder } from './TradeOrder';
import { Client } from './Client';

export class RuleContext {
  readonly order: TradeOrder;
  readonly client: Client | null;
  private readonly attributes: Map<string, unknown> = new Map();
  private readonly _messages: string[] = [];
  private readonly _warnings: string[] = [];
  private _rejected = false;
  private _rejectionReason = '';

  constructor(order: TradeOrder, client: Client | null) {
    this.order = order;
    this.client = client;
  }

  getAttribute(key: string): unknown {
    return this.attributes.get(key);
  }

  setAttribute(key: string, value: unknown): void {
    this.attributes.set(key, value);
  }

  get messages(): string[] {
    return [...this._messages];
  }

  addMessage(msg: string): void {
    this._messages.push(msg);
  }

  get warnings(): string[] {
    return [...this._warnings];
  }

  addWarning(warning: string): void {
    this._warnings.push(warning);
  }

  get rejected(): boolean {
    return this._rejected;
  }

  get rejectionReason(): string {
    return this._rejectionReason;
  }

  reject(reason: string): void {
    this._rejected = true;
    this._rejectionReason = reason;
  }
}
